### PR TITLE
docs: Update DG NFRs and Use cases

### PR DESCRIPTION
**Pull request by HugeNoob**
## What type of PR is this? (check all applicable)

-   [ ] Refactor
-   [ ] Feature
-   [ ] Bug Fix
-   [ ] Optimization
-   [x] Documentation Update

## Description

- Add NFRs (many referenced from [here](https://nus-cs2103-ay2324s1.github.io/website/admin/tp-constraints.html#tp-constraints)).
- Add use cases for:
  - `help`
  - `deleteAllPerson`
  - `editTask`
  - `deleteTask`

## Related Tickets & Documents

-   Related Issue https://github.com/AY2324S1-CS2103T-T10-2/tp/issues/39

## Added/updated tests?

_We encourage you to keep the code coverage percentage at 75% and above for new changes._

-   [x] No, and this is why: Deployment on [branch](https://docs-dg-v1-1.coordimate.pages.dev/DeveloperGuide).

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -308,15 +311,95 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+2. CoordiMate shows a confirmation message.
+
+    Use case ends.
+
+**Extensions**
+
+* 2a. User requests to list all contacts, to verify that all contacts are deleted.
+    * 2a1. CoordiMate shows an empty contact list.
+
+      Use case ends.
```
Comment (jiakai-17) [inline docs/DeveloperGuide.md:338]: I dont believe that this extension is necessary, as the listPersons use case would have already covered the case where the contact list is empty.

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -308,15 +311,95 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+* 2a. The list is empty.
+
+  Use case ends.
+
+* 3a. The given index is invalid.
+
+    * 3b1. CoordiMate shows an error message and prompts the user to enter a valid index.
+    * 3b2. User requests a deletion with a valid index.
+
+      Use case resumes at step 4.
```
Comment (jiakai-17) [inline docs/DeveloperGuide.md:387]: Perhaps you could correct this numbering? 

Also, I think the use case should redirect back to step 3, as the user would need to enter the whole command again.

```suggestion
    * 3a1. CoordiMate shows an error message and prompts the user to enter a valid index.

      Use case resumes from step 3.
```

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -308,15 +311,95 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+* 2a. The list is empty.
+
+  Use case ends.
+
+* 3a. The given index is invalid.
+
+    * 3b1. CoordiMate shows an error message and prompts the user to enter a valid index.
+    * 3b2. User requests an edit with a valid index.
+
+      Use case resumes at step 4.
```
Comment (jiakai-17) [inline docs/DeveloperGuide.md:363]: Perhaps you could correct this numbering? 

Also, I think the use case should redirect back to step 3, as the user would need to enter the whole command again.

```suggestion
    * 3a1. CoordiMate shows an error message and prompts the user to enter a valid index.

      Use case resumes from step 3.
```

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -308,15 +311,95 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
 
       Use case resumes at step 2.
 
+**Use case: UC01 - See help instructions**
+
+**MSS**
+
+1. User requests for help.
+2. AddressBook shows the help instructions, with a link to user guide.
```
Comment (jiakai-17) [inline docs/DeveloperGuide.md:319]:
```suggestion
2. CoordiMate shows the help instructions, with a link to user guide.
```

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -272,9 +272,12 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
 | -------- | ------------------------------------------ | ------------------------------ | ---------------------------------------------------------------------- |
-| `* * *`  | new user                                   | see usage instructions         | refer to instructions when I forget how to use the App                 |
+| `* * *`  | new user                                   | see help instructions          | refer to documentation to understand the existing features effectively |
 | `* * *`  | user                                       | add a new person               |                                                                        |
 | `* * *`  | user                                       | delete a person                | remove entries that I no longer need                                   |
+| `* * *`  | event planner                              | delete all contacts            | efficiently restart or declutter my addressbook                        |
+| `* * *`  | event planner                              | editTask                       | ensure task details are up-to-date with latest information             |
```
Comment (jiakai-17) [inline docs/DeveloperGuide.md:279]:
```suggestion
| `* * *`  | event planner                              | edit a task                    | ensure task details are up-to-date with latest information             |
```

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -272,9 +272,12 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+| `* * *`  | event planner                              | deleteTask                     | remove tasks that are no longer relevant                               |
```
Comment (jiakai-17) [inline docs/DeveloperGuide.md:280]:
```suggestion
| `* * *`  | event planner                              | delete a task                  | remove tasks that are no longer relevant                               |
```

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -308,15 +311,95 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+
+    Use case ends.
+
+**Use case: UC07 - Delete all existing contacts**
+
+**MSS**
+
+1. User requests to delete all contacts.
+2. CoordiMate shows a confirmation message.
+
```
Comment (jiakai-17) [inline docs/DeveloperGuide.md:330]: Perhaps you could add a step about the user accepts the confirmation message and another step about CoordiMate deleting all contacts.

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -308,15 +311,95 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+    * 3b2. User requests a deletion with a valid index.
+
+      Use case resumes at step 4.
+
 *{More to be added}*
 
 ### Non-Functional Requirements
 
 1.  Should work on any _mainstream OS_ as long as it has Java `11` or above installed.
 2.  Should be able to hold up to 1000 persons without a noticeable sluggishness in performance for typical usage.
```
Comment (jiakai-17) [inline docs/DeveloperGuide.md:394]: Perhaps you could modify this NFR to include tasks as well?

```suggestion
2.  Should be able to hold up to 1000 persons and tasks without a noticeable sluggishness in performance for typical usage.
```

**Review by jiakai-17 — CHANGES_REQUESTED** [review]
Thanks for the changes, just some minor nit about the content and formatting.

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (HugeNoob) [reply docs/DeveloperGuide.md:338]: It's more of an additional step to check that the tasks are deleted properly. I imagine many users would want to do this. I'm not sure what's worth adding in as an extension, wdyt?

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (HugeNoob) [reply docs/DeveloperGuide.md:319]: Copilot moment

File: docs/DeveloperGuide.md (hunk 7 above)
Comment (HugeNoob) [reply docs/DeveloperGuide.md:330]: Is this an intended feature? The confirmation message here is more of a "receipt" after the deletion, rather than a confirmation prompt.

Comment (jiakai-17) [reply docs/DeveloperGuide.md:330]: In that case, could you rephrase it to be more clear?
For example:

2. CoordiMate deletes all contacts and shows a success message.

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (HugeNoob) [reply docs/DeveloperGuide.md:338]: Decided to remove after discussion.

Comment (m1oojv) [reply docs/DeveloperGuide.md]: "Extensions are "add-on"s to the MSS that describe exceptional/alternative flow of events. They describe variations of the scenario that can happen if certain things are not as expected by the MSS. Extensions appear below the MSS." 
- from CS2103T website

So i think verifying is not really an alternative flow of events (more like an after flow). I think it'll be better to rephrase the extension to 
The list is empty . Use case ends 

But feel free to disagree - Wdyt

**Review by jiakai-17 — APPROVED** [review]
